### PR TITLE
Implement FinalizationRegistry for JNI platforms

### PR DESCRIPTION
**Pull request by swankjesse**

### CODE REVIEW

**Comment by swankjesse** [comment]
Next steps on this so it also works on iOS:

 * [x] Change the opaque object on Finalizer from `int` to a `Finalizer` struct containing the `JSContext*`, `finalizerClassId`, and `id`. Heap-allocate that thing.
 * [x] Drop the dependency from `FinalizationRegistry.cpp` onto JNI and `Context.cpp`.
 * [x] Convert `FinalizationRegistry.cpp` to C
 * [x] Move `FinalizationRegistry.cpp` to `zipline/native/common`.

**Review by JakeWharton — APPROVED** [review]

**Comment by JakeWharton** [comment]
I guess all the work has to happen before it lands? Or are you going to move the tests to jniTest so it can land?

**Comment by swankjesse** [comment]
I wanna see how much I can get done today. If I’m lucky I’ll just push another commit that gets it all working?

**Comment by swankjesse** [comment]
@JakeWharton all four follow-ups added to the PR. Mind taking another look?

File: zipline/native/common/finalization-registry.c [hunk 1]
```diff
@@ -36,37 +33,41 @@
...
+} FinalizerOpaque;
 
 /*
  * This is invoked by QuickJS when an instance of our special Finalizer class is garbage collected.
  * It is equivalent to the following pseudocode.
  *
  * function jsFinalizerCollected(value) {
- *   const id = value.magicOpaqueValue;
- *   globalThis.app_cash_zipline_enqueueFinalizer(id);
+ *   val finalizerOpaque = value.magicOpaqueValue;
```
Comment (JakeWharton) [inline zipline/native/common/finalization-registry.c]:
```suggestion
 *   const finalizerOpaque = value.magicOpaqueValue;
```

**Review by JakeWharton — APPROVED** [review]

File: zipline/native/common/finalization-registry.c (hunk 1 above)
Comment (swankjesse) [reply zipline/native/common/finalization-registry.c]: Fixed.

**Comment by swankjesse** [comment]
I must admit this fix took a couple hours longer than I would have liked.
https://github.com/cashapp/zipline/pull/486/commits/d62b0e4ca6e027ce3a92b71bd7d8e3e96fdb0c3a

**Pull request merged**